polaris-service/src/main/java/org/apache/polaris/service/catalog/io/WasbTranslatingFileIO.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.catalog.io;
+
+import java.util.Map;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.polaris.core.storage.StorageLocation;
+import org.apache.polaris.core.storage.azure.AzureLocation;
+
+/**
+ * A {@link FileIO} implementation that translates WASB paths into ABFS paths and then delegates to
+ * another underlying FileIO implementation
+ */
+public class WasbTranslatingFileIO implements FileIO {
+  private final FileIO io;
+
+  private static final String WASB_SCHEME = "wasb";
+  private static final String ABFS_SCHEME = "abfs";
+
+  public WasbTranslatingFileIO(FileIO io) {
+    this.io = io;
+  }
+
+  private static String translate(String path) {
+    if (path == null) {
+      return null;
+    } else {
+      StorageLocation storageLocation = StorageLocation.of(path);
+      if (storageLocation instanceof AzureLocation azureLocation) {
+        String scheme = azureLocation.getScheme();
+        if (scheme.startsWith(WASB_SCHEME)) {
+          scheme = scheme.replaceAll(WASB_SCHEME, ABFS_SCHEME);
```

Comment:
Though it *shouldn't* matter, `replaceFirst` might convey intent a bit better, but ideally we can just be more explicit with a `Map<String, String>` lookup anyways -- being explicit about both `wasb -> abfs` and `wasbs -> abfss` in this case is probably worth the extra boilerplate even if we don't get to "save" on handling the "optional" "s" suffix.